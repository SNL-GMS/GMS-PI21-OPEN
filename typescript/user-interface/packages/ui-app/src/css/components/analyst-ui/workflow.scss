@import '../../gms-fonts.scss';

$BLOCK_HEIGHT_PX: 24px;
$TABLE_BUTTON_WIDTH: 24px;
$TABLE_LABEL_WIDTH: 137px;
$WIDTH_OF_LEFT_CURTAIN: 24px;
$WIDTH_OF_RIGHT_CURTAIN: 140px;
$WIDTH_OF_RIGHT_PADDING: 148px;

.workflow-display-window {
  height: 100%;
  width: 100%;
  overflow: hidden;
  background-color: var(--gms-background, $gms-background);
}

.workflow-panel {
  height: 100%;
  width: 100%;
}

.workflow-table {
  display: flex;
  flex-direction: column;
}

.workflow-table-container {
  display: flex;
  position: relative;
  overflow-y: scroll;
  overflow-x: hidden;
  gap: 2px;
}

.workflow-table__curtain-left {
  width: $WIDTH_OF_LEFT_CURTAIN;
  position: absolute;
  background-color: var(--gms-background, $gms-background);
  z-index: 2;
}

.workflow-table__curtain-right {
  width: $WIDTH_OF_RIGHT_CURTAIN;
  position: absolute;
  right: 0px;
  background-color: var(--gms-background, $gms-background);
  z-index: 2;
  top: 0px;
}

.workflow-table__buttons {
  flex-direction: column;
  width: $TABLE_BUTTON_WIDTH;
  display: flex;
  align-items: flex-start;
}

.workflow-button {
  width: $TABLE_BUTTON_WIDTH;
}

.workflow-button__blank {
  height: $BLOCK_HEIGHT_PX;
  min-height: $BLOCK_HEIGHT_PX;
  z-index: 5;
  width: $TABLE_BUTTON_WIDTH;
}

.workflow-button__expand_collapse {
  height: $BLOCK_HEIGHT_PX;
  border-color: var(--gms-background, $gms-background);
  transition: margin-bottom 0.15s;
  z-index: 5;
  width: $TABLE_BUTTON_WIDTH;
}

.workflow-table__labels {
  display: flex;
  flex-direction: column;
}

.workflow-table-label {
  height: $BLOCK_HEIGHT_PX;
  width: $TABLE_LABEL_WIDTH;
  display: flex;
  justify-content: center;
  align-items: center;
  background: var(--gms-soft, $gms-soft);
  z-index: 5;
  -webkit-box-sizing: border-box;
  -moz-box-sizing: border-box;
  box-sizing: border-box;
  border: thin solid var(--gms-background, $gms-background);
}

.workflow-table-label--activity {
  filter: $interval-activity-darken-cell-by;
}

.workflow-table__time-axis {
  font-family: $gms-mono;
  margin-left: $TABLE_BUTTON_WIDTH;
  position: relative;
}

.time-axis-wrapper {
  width: 100%;
  height: 25px;
  flex: 0 0 auto;
  position: static;
  z-index: 1;
}

.scroll-bar-override {
  width: 100%;
  height: 14px;
  flex: 0 0 auto;
  position: static;
  z-index: 1;
}

.workflow-time-axis {
  path,
  line {
    fill: none;
    stroke: var(--gms-prominent, $gms-prominent);
    shape-rendering: crispEdges;
  }
  text {
    font-family: $gms-mono;
    font-size: 12px;
    fill: var(--gms-main, $gms-main);
    &.day-label {
      stroke: $workflow-day-divider;
      font-size: 13px;
    }
    &.first {
      text-anchor: start;
    }
    &.last {
      text-anchor: end;
    }
    &.hidden {
      display: none;
    }
  }
}

.time-axis {
  width: 100%;
  height: 25px;
  overflow: hidden;
}

.workflow-time-range {
  font-size: 0.75em;
  position: absolute;
  width: 100%;
  line-height: 1;
}

.time-range-left {
  float: left;
  line-height: 1;
}

.time-range-right {
  float: right;
  line-height: 1;
  margin-right: $WIDTH_OF_RIGHT_PADDING;
}

.stage-row__expand-button {
  height: $BLOCK_HEIGHT_PX;
  border: thin solid var(--gms-recessed, $gms-recessed);
  width: $BLOCK_HEIGHT_PX;
  transition: margin-bottom 0.15s;
  min-height: $BLOCK_HEIGHT_PX;
  min-width: $BLOCK_HEIGHT_PX;
  padding: 0;
  position: absolute;
  z-index: 5;
  -webkit-box-sizing: border-box;
  -moz-box-sizing: border-box;
  box-sizing: border-box;
}

.workflow-tooltip-container {
  display: grid;
  grid-template-columns: 2fr 3fr;
  margin: 0 auto;
  gap: 0.25em;
  padding: 0 0.5em;
  background-color: var(--gms-grid-background, $gms-grid-background);
}

.workflow-tooltip-container .label-value__value {
  color: var(--gms-main, $gms-main);
  font-size: 14px;
  &--numeric {
    font-family: $gms-mono;
  }
}

.workflow-tooltip-container .label-value__label {
  font-size: 14px;
}

.workflow-tooltip-container:last-child {
  padding-bottom: 0;
}

.workflow-tooltip-container + .workflow-tooltip-container {
  margin-top: 3px;
}

.interval-cell {
  display: flex;
  height: $BLOCK_HEIGHT_PX;
  align-items: center;
  text-align: center;
  justify-content: center;
  margin: 0;
  overflow: hidden;
  white-space: nowrap;
  .percent-bar {
    height: 99%;
    width: 99%;
  }
  -webkit-box-sizing: border-box;
  -moz-box-sizing: border-box;
  box-sizing: border-box;
  border: thin solid var(--gms-background, $gms-background);
}

.interval-cell:hover {
  filter: brightness(1.2);
}

.interval-cell--in-progress {
  background-color: $interval-in-progress-fill;
  .percent-bar {
    background-color: scale-color(
      $color: $interval-in-progress-fill,
      $lightness: 25%,
      $saturation: -15%
    );
  }
}

.interval-cell--complete {
  background-color: $interval-complete-fill;
}

.interval-cell--not-complete {
  background-color: $interval-not-complete-fill;
}

.interval-cell--not-started {
  background-color: $interval-not-started-fill;
}

.interval-cell--skipped {
  background-color: $interval-skipped-fill;
}

.interval-cell--failed {
  background-color: $interval-failed-fill;
}

.interval-cell--selected {
  border: 2px solid $gms-selection;
}

.interval-cell--activity-cell {
  filter: $interval-activity-darken-cell-by;
}

.interval-cell--activity-cell:hover {
  filter: brightness(1);
}

.interval-cell--stale {
  filter: brightness(0.5);
}

.interval-cell--clickable {
  cursor: pointer;
}

.workflow-day-boundary-container {
  position: absolute;
}

.workflow-day-divider {
  position: absolute;
  top: 0px;
  bottom: 0px;
  width: 2px;
  z-index: 1;
  margin-left: -1px;
  background-color: #ffc940;
}

.workflow-ellipsis {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  width: 100px;
  text-align: center;
}

.workflow-open-anything-row {
  display: flex;
  flex-flow: wrap;
  box-sizing: border-box;
  align-items: center;
  min-width: 56ch;
  & > * {
    margin-left: 0.5rem;
  }
}

.open-anything-contents .dropdown-container {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-top: 7px;
}

.workflow-open-anything-row .dropdown-container .span {
  display: flex;
}

.open-anything-contents .dropdown-label {
  padding: 0 0.5em 0 0;
  line-height: 1;
  color: var(--core-prominent, var(--gms-prominent, $core-prominent));
  font-weight: 600;
  width: 35%;
}

.open-anything-contents .dropdown-selector {
  width: 65%;
}

.interval-confirmation-contents {
  display: flex;
}

.interval-confirmation-text {
  width: 90%;
}

.interval-confirmation-icon {
  width: 10%;
}

.interval-confirmation-header {
  font-weight: bold;
}

.open-anything-error {
  color: $gms-strong-warning;
  display: flex;
  min-height: 20px;
}
.open-anything-error-text {
  margin-left: 5px;
  width: calc(100% - 20px);
}
