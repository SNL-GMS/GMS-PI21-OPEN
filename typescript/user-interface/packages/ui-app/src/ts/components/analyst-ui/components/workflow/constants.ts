import { SECONDS_IN_MINUTES } from '@gms/common-util';

export const TABLE_SIDE_PADDING = 20;
export const TABLE_BUTTON_WIDTH = 32;
export const TABLE_LABEL_WIDTH = 135;
export const TOOLBAR_HEIGHT_PX = 35;
export const TIME_AXIS_HEIGHT = 35;

export const HUNDRED_TWENTY_SECOND_BUFFER = 120;
export const HEIGHT = 24;
export const PIXELS_PER_SECOND = 2 / SECONDS_IN_MINUTES;
