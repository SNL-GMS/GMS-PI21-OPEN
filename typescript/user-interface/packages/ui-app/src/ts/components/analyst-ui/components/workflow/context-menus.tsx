import { Menu, MenuItem } from '@blueprintjs/core';
import type { WorkflowTypes } from '@gms/common-model';
import { useAppSelector } from '@gms/ui-state';
import React from 'react';

export interface IntervalContextMenuProps {
  readonly interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval;
  readonly isSelectedInterval: boolean;
  readonly allActivitiesOpenForSelectedInterval: boolean;

  readonly openCallback: (
    interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval
  ) => void;
  readonly closeCallback: (
    interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval
  ) => void;
}

/**
 * Component that renders the interval context menu.
 */
// eslint-disable-next-line react/function-component-definition
export const IntervalContextMenu: React.FunctionComponent<IntervalContextMenuProps> = (
  props: IntervalContextMenuProps
) => {
  const openIntervalName = useAppSelector(state => state.app.workflow.openIntervalName);
  const {
    interval,
    isSelectedInterval,
    allActivitiesOpenForSelectedInterval,
    openCallback,
    closeCallback
  } = props;

  const isDisabled =
    (allActivitiesOpenForSelectedInterval &&
      isSelectedInterval &&
      openIntervalName === interval.name) ||
    (isSelectedInterval && interval.name !== openIntervalName);
  return (
    <Menu>
      <MenuItem
        className="menu-item-open-interval"
        data-cy="open-interval-btn"
        text="Open interval"
        disabled={isDisabled}
        onClick={() => openCallback(interval)}
      />
      <MenuItem
        className="menu-item-close-interval"
        data-cy="close-interval-btn"
        text="Close interval"
        disabled={!isSelectedInterval}
        onClick={() => closeCallback(interval)}
      />
    </Menu>
  );
};
