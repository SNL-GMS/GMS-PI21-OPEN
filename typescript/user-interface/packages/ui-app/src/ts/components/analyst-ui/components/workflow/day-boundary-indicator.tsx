import type { CommonTypes } from '@gms/common-model';
import { MILLISECONDS_IN_SECOND, toDate } from '@gms/common-util';
import { UILogger } from '@gms/ui-util';
import * as d3 from 'd3';
import React from 'react';

import { getScaleForTimeRange } from './workflow-util';

const logger = UILogger.create('GMS_LOG_WORKFLOW', process.env.GMS_LOG_WORKFLOW);

/**
 * DayBoundaryIndicator Props
 */
export interface DayBoundaryIndicatorProps {
  readonly timeRange: CommonTypes.TimeRange;
  readonly width: number;
  readonly height: number;
}

/**
 * @DayBoundaryIndicator
 * The marker between days in the workflow table
 * It is kept in synch with the workflow table through its onScroll callback
 * Line indicating the transition from one day from the next
 */
export class DayBoundaryIndicator extends React.Component<DayBoundaryIndicatorProps> {
  private dayBoundaryContainerRef: HTMLElement;

  public shouldComponentUpdate(nextProps: DayBoundaryIndicatorProps): boolean {
    return this.internalUpdate(nextProps);
  }

  private readonly internalUpdate = (
    nextProps: DayBoundaryIndicatorProps,
    force = false
  ): boolean => {
    const {
      height,
      timeRange: { startTimeSecs, endTimeSecs }
    } = this.props;
    const { height: nextHeight, timeRange: nextTimeRange } = nextProps;
    const { startTimeSecs: nextStartTimeSecs, endTimeSecs: nextEndTimeSecs } = nextTimeRange;

    const shouldUpdate =
      force || startTimeSecs !== nextStartTimeSecs || endTimeSecs !== nextEndTimeSecs;

    // no reason to update the height; if about to rerender
    if (!shouldUpdate && height !== nextHeight) {
      logger.debug(`Updating the DayBoundaryIndicator height`, height, nextHeight);
      this.dayBoundaryContainerRef.style.height = `${nextHeight}px`;
    }
    return shouldUpdate;
  };

  /**
   * Synchronizes the day indicator with the main display
   *
   * @param scrollTo pixel value reflecting how scrolled the workflow is
   */
  public readonly scrollDayIndicator = (scrollTo: number): void => {
    if (!this.dayBoundaryContainerRef) return;
    this.dayBoundaryContainerRef.style.marginLeft = `${(-scrollTo).toString()}px`;
  };

  public render(): JSX.Element {
    const { height, timeRange } = this.props;
    const { totalWidth, scaleToPosition } = getScaleForTimeRange(timeRange);

    const days = d3.utcDay
      .every(1)
      .range(toDate(timeRange.startTimeSecs), toDate(timeRange.endTimeSecs));

    logger.debug(`Rendering DayBoundaryIndicator`, this.props, days.length);

    return (
      <div
        ref={ref => {
          this.dayBoundaryContainerRef = ref;
        }}
        className="workflow-day-boundary-container"
        style={{ width: totalWidth, height }}
      >
        {days.map(value => {
          return (
            <div
              key={value.valueOf()}
              className="workflow-day-divider"
              style={{
                left: `${scaleToPosition(value.valueOf() / MILLISECONDS_IN_SECOND)}px`,
                height: `100%`
              }}
            />
          );
        })}
      </div>
    );
  }
}
