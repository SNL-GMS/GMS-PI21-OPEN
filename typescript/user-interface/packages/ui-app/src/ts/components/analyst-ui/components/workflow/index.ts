import { WorkflowComponent as Workflow } from './workflow-component';
import * as WorkflowUtil from './workflow-util';

export { Workflow, WorkflowUtil };
