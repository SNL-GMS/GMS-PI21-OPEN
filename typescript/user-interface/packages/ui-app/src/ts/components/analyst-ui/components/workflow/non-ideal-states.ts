import type { NonIdealStateDefinition } from '@gms/ui-core-components';
import { nonIdealStateWithNoSpinner, nonIdealStateWithSpinner } from '@gms/ui-core-components';
import type { StageIntervalsByIdAndTimeQuery, WorkflowQuery } from '@gms/ui-state';

export const workflowQueryNonIdealStates: NonIdealStateDefinition<unknown>[] = [
  {
    condition: (props: { workflowQuery: WorkflowQuery }): boolean => {
      return props.workflowQuery?.isLoading;
    },
    element: nonIdealStateWithSpinner('Loading', 'Workflow Data')
  },
  {
    condition: (props: { workflowQuery: WorkflowQuery }): boolean => {
      return props.workflowQuery?.isError;
    },
    element: nonIdealStateWithSpinner('Error', 'Problem Loading Workflow Data')
  }
];

export const workflowIntervalQueryNonIdealStates: NonIdealStateDefinition<unknown>[] = [
  {
    condition: (props: {
      readonly hasFetchedInitialIntervals: boolean;
      workflowIntervalQuery: StageIntervalsByIdAndTimeQuery;
    }): boolean => {
      // only show the loading state if the on the initial fetch of the interval data
      return !props.hasFetchedInitialIntervals && props.workflowIntervalQuery?.isLoading;
    },
    element: nonIdealStateWithSpinner('Loading', 'Workflow Interval Data')
  },
  {
    condition: (props: { workflowIntervalQuery: StageIntervalsByIdAndTimeQuery }): boolean => {
      return (
        !props.workflowIntervalQuery?.isLoading &&
        !props.workflowIntervalQuery?.isError &&
        (!props.workflowIntervalQuery.data || props.workflowIntervalQuery.data.length === 0)
      );
    },
    element: nonIdealStateWithNoSpinner('No data', 'Stage intervals returned empty')
  },
  {
    condition: (props: { workflowIntervalQuery: StageIntervalsByIdAndTimeQuery }): boolean => {
      return props.workflowIntervalQuery?.isError;
    },
    element: nonIdealStateWithNoSpinner('Error', 'Problem Loading Workflow Interval Data')
  }
];
