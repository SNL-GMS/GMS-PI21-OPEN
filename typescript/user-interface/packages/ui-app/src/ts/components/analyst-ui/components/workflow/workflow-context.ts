import type { WorkflowTypes } from '@gms/common-model';
import * as React from 'react';

import type { OpenAnythingInterval } from './types';

export interface WorkflowContextData {
  readonly staleStartTime: number;
  readonly allActivitiesOpenForSelectedInterval: boolean;
  readonly closeConfirmationPrompt: (
    interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval
  ) => void;
  readonly openConfirmationPrompt: (
    interval: WorkflowTypes.ActivityInterval | WorkflowTypes.StageInterval
  ) => void;
  readonly openAnythingConfirmationPrompt: (interval: OpenAnythingInterval) => void;
}

/**
 * The audible notification context
 */
export const WorkflowContext: React.Context<WorkflowContextData> = React.createContext<
  WorkflowContextData
>(undefined);
