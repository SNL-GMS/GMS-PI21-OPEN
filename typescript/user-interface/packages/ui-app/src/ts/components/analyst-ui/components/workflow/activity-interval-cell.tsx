import { ContextMenu } from '@blueprintjs/core';
import { WorkflowTypes } from '@gms/common-model';
import { useAppSelector } from '@gms/ui-state';
import classNames from 'classnames';
import includes from 'lodash/includes';
import React from 'react';
import { Provider, useStore } from 'react-redux';

import { IntervalContextMenu } from './context-menus';
import { WorkflowContext } from './workflow-context';

export interface ActivityIntervalCellProps {
  readonly activityInterval: WorkflowTypes.ActivityInterval;
}

/**
 * Determines the text for the cell based on status and active analyst list
 *
 * @param status interval status
 * @param activeAnalysts list of activeAnalyst
 * @returns text for cell to display
 */
export const determineTextForCell = (
  status: WorkflowTypes.IntervalStatus,
  activeAnalysts: string[]
): string => {
  if (!activeAnalysts || activeAnalysts.length === 0) return '';

  switch (status) {
    case WorkflowTypes.IntervalStatus.NOT_STARTED:
    case WorkflowTypes.IntervalStatus.NOT_COMPLETE:
      // !note for complete the name needs to be analyst that marked complete
      return '';
    case WorkflowTypes.IntervalStatus.COMPLETE:
      return activeAnalysts[0];
    default:
      return `${activeAnalysts[0]}${
        activeAnalysts.length > 1 ? ` + ${activeAnalysts.length - 1}` : ''
      }`;
  }
};

export const preventDefaultEvent = (event: React.MouseEvent<HTMLDivElement, MouseEvent>): void => {
  event.preventDefault();
};

// eslint-disable-next-line react/display-name
export const ActivityIntervalCell: React.FunctionComponent<ActivityIntervalCellProps> = React.memo(
  (props: ActivityIntervalCellProps) => {
    const { activityInterval } = props;
    const openIntervalName = useAppSelector(state => state.app.workflow.openIntervalName);
    const openActivityNames = useAppSelector(state => state.app.workflow.openActivityNames);
    const startTimeSecs = useAppSelector(state => state.app.workflow.timeRange.startTimeSecs);
    const context = React.useContext(WorkflowContext);
    const isSelected =
      openIntervalName === activityInterval.stageName &&
      includes(openActivityNames, activityInterval.name) &&
      startTimeSecs === activityInterval.startTime;
    const isStale = context.staleStartTime >= activityInterval.endTime;
    const isClickable = !isStale || isSelected;
    const cellClass = classNames({
      'interval-cell': true,
      'interval-cell--selected': isSelected,
      'interval-cell--not-complete':
        activityInterval.status === WorkflowTypes.IntervalStatus.NOT_COMPLETE,
      'interval-cell--in-progress':
        activityInterval.status === WorkflowTypes.IntervalStatus.IN_PROGRESS,
      'interval-cell--not-started':
        activityInterval.status === WorkflowTypes.IntervalStatus.NOT_STARTED,
      'interval-cell--complete': activityInterval.status === WorkflowTypes.IntervalStatus.COMPLETE,
      'interval-cell--activity-cell': true,
      'interval-cell--stale': isStale,
      'interval-cell--clickable': isClickable
    });
    const store = useStore();
    return (
      <div
        key={activityInterval.startTime}
        data-cy={`${activityInterval.startTime}-${activityInterval.name}`}
        data-start-time={activityInterval.startTime}
        data-activity-interval={`${activityInterval.name}`}
        className={cellClass}
        onDoubleClick={() => {
          if (!isStale && !isSelected) {
            context.openConfirmationPrompt(activityInterval);
          }
        }}
        // if stale do not allow context menu, unless that interval is already selected (active)
        onContextMenu={
          isClickable
            ? e => {
                e.preventDefault();
                // TODO update to ContextMenu2 see https://blueprintjs.com/docs/#popover2-package/context-menu2
                ContextMenu.show(
                  <Provider store={store}>
                    <IntervalContextMenu
                      interval={activityInterval}
                      isSelectedInterval={isSelected}
                      allActivitiesOpenForSelectedInterval={
                        context.allActivitiesOpenForSelectedInterval
                      }
                      openCallback={context.openConfirmationPrompt}
                      closeCallback={context.closeConfirmationPrompt}
                    />
                  </Provider>,
                  {
                    left: e.clientX,
                    top: e.clientY
                  },
                  undefined,
                  true
                );
              }
            : preventDefaultEvent
        }
      >
        <span className="workflow-ellipsis">
          {determineTextForCell(activityInterval.status, activityInterval.activeAnalysts)}
        </span>
      </div>
    );
  }
);
