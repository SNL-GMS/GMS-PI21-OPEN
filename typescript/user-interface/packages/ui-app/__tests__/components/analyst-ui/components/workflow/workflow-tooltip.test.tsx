/* eslint-disable react/jsx-no-constructed-context-values */
/* eslint-disable react/jsx-props-no-spreading */
import { WorkflowTypes } from '@gms/common-model';
import { secondsToString } from '@gms/common-util';
import { render } from '@testing-library/react';
import * as React from 'react';

import { WorkflowContext } from '../../../../../src/ts/components/analyst-ui/components/workflow/workflow-context';
import {
  getActiveAnalysts,
  getStatus,
  Tooltip,
  TooltipPanel
} from '../../../../../src/ts/components/analyst-ui/components/workflow/workflow-tooltip';
import * as WorkFlowDataTypes from './workflow-data-types';

// set up window alert and open so we don't see errors
window.alert = jest.fn();
window.open = jest.fn();

window.ResizeObserver = jest.fn(() => {
  return { observe: jest.fn(), disconnect: jest.fn(), unobserve: jest.fn() };
});
const globalAny: any = global;
globalAny.ResizeObserver = window.ResizeObserver;
globalAny.DOMRect = jest.fn(() => ({}));

const MOCK_TIME = 1606818240000;
global.Date.now = jest.fn(() => MOCK_TIME);
describe('Workflow Tooltip', () => {
  it('is exported', () => {
    expect(TooltipPanel).toBeDefined();
    expect(getActiveAnalysts).toBeDefined();
    expect(Tooltip).toBeDefined();
  });

  it('can get active analysts', () => {
    expect(getActiveAnalysts(undefined)).toMatchInlineSnapshot(`undefined`);

    expect(
      getActiveAnalysts(WorkFlowDataTypes.interactiveAnalysisStageInterval)
    ).toMatchInlineSnapshot(`undefined`);

    expect(getActiveAnalysts(WorkFlowDataTypes.activityInterval)).toMatchInlineSnapshot(
      `"larry, moe, curly"`
    );
  });

  it('can get status', () => {
    expect(getStatus).toBeDefined();

    expect(getStatus(undefined)).toMatchInlineSnapshot(`undefined`);

    expect(getStatus(WorkFlowDataTypes.automaticProcessingStageInterval)).toMatchInlineSnapshot(
      `"In Progress (last-step)"`
    );

    expect(getStatus(WorkFlowDataTypes.activityInterval)).toMatchInlineSnapshot(`"In Progress"`);

    expect(getStatus(WorkFlowDataTypes.automaticProcessingStageInterval)).toMatchInlineSnapshot(
      `"In Progress (last-step)"`
    );

    expect(getStatus(WorkFlowDataTypes.activityInterval)).toMatchInlineSnapshot(`"In Progress"`);

    expect(
      getStatus(
        (WorkFlowDataTypes.processingSequenceInterval as unknown) as WorkflowTypes.StageInterval
      )
    ).toMatchInlineSnapshot(`"In Progress (last-step)"`);
  });

  it('matches tooltip panel snapshot', () => {
    expect(
      render(
        <WorkflowContext.Provider
          value={{
            staleStartTime: 1,
            allActivitiesOpenForSelectedInterval: false,
            openConfirmationPrompt: jest.fn(),
            openAnythingConfirmationPrompt: jest.fn(),
            closeConfirmationPrompt: jest.fn()
          }}
        >
          <TooltipPanel
            startTime={undefined}
            endTime={undefined}
            status={undefined}
            activeAnalysts={undefined}
            lastModified={undefined}
            setTooltipKey={jest.fn()}
            tooltipRef={{
              current: {
                focus: jest.fn()
              } as any
            }}
            isStale={false}
          />
        </WorkflowContext.Provider>
      ).container
    ).toMatchSnapshot();

    expect(
      render(
        <WorkflowContext.Provider
          value={{
            staleStartTime: 1,
            allActivitiesOpenForSelectedInterval: false,
            openConfirmationPrompt: jest.fn(),
            openAnythingConfirmationPrompt: jest.fn(),
            closeConfirmationPrompt: jest.fn()
          }}
        >
          <TooltipPanel
            startTime="500"
            endTime="600"
            status={WorkflowTypes.IntervalStatus.IN_PROGRESS}
            activeAnalysts="analyst 1, analyst 2"
            lastModified={secondsToString(0)}
            setTooltipKey={undefined}
            tooltipRef={{
              current: {
                focus: jest.fn()
              } as any
            }}
            isStale={false}
          />
        </WorkflowContext.Provider>
      ).container
    ).toMatchSnapshot();
  });

  it('matches tooltip snapshot', () => {
    expect(
      render(
        <WorkflowContext.Provider
          value={{
            staleStartTime: 1,
            allActivitiesOpenForSelectedInterval: false,
            openConfirmationPrompt: jest.fn(),
            openAnythingConfirmationPrompt: jest.fn(),
            closeConfirmationPrompt: jest.fn()
          }}
        >
          <Tooltip interval={WorkFlowDataTypes.activityInterval} staleStartTime={500} />
        </WorkflowContext.Provider>
      ).container
    ).toMatchSnapshot();

    expect(
      render(
        <WorkflowContext.Provider
          value={{
            staleStartTime: 1,
            allActivitiesOpenForSelectedInterval: false,
            openConfirmationPrompt: jest.fn(),
            openAnythingConfirmationPrompt: jest.fn(),
            closeConfirmationPrompt: jest.fn()
          }}
        >
          <Tooltip
            interval={WorkFlowDataTypes.automaticProcessingStageInterval}
            staleStartTime={500}
          >
            <div>content</div>
          </Tooltip>
        </WorkflowContext.Provider>
      ).container
    ).toMatchSnapshot();
  });

  it('can handle undefined interval with tooltip', () => {
    expect(
      render(
        <WorkflowContext.Provider
          value={{
            staleStartTime: 1,
            allActivitiesOpenForSelectedInterval: false,
            openConfirmationPrompt: jest.fn(),
            openAnythingConfirmationPrompt: jest.fn(),
            closeConfirmationPrompt: jest.fn()
          }}
        >
          <Tooltip interval={undefined} staleStartTime={500} />{' '}
        </WorkflowContext.Provider>
      ).container
    ).toMatchSnapshot();
  });
});
