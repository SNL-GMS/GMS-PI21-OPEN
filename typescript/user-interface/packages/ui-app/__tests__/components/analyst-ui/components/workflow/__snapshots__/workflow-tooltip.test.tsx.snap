// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workflow Tooltip can handle undefined interval with tooltip 1`] = `
<div>
   
</div>
`;

exports[`Workflow Tooltip matches tooltip panel snapshot 1`] = `
<div>
  <div
    class="workflow-tooltip"
    data-cy="workflow-tooltip"
    tabindex="0"
  />
</div>
`;

exports[`Workflow Tooltip matches tooltip panel snapshot 2`] = `
<div>
  <div
    class="workflow-tooltip"
    data-cy="workflow-tooltip"
    tabindex="0"
  >
    <div
      class="label-value-container workflow-tooltip-container"
    >
      <div
        class="label-value__label"
        data-cy="status-label"
      >
        Status: 
      </div>
      <div
        class="label-value__value"
        data-cy="status-value"
        title="Status"
      >
        IN_PROGRESS 
      </div>
    </div>
    <div
      class="label-value-container workflow-tooltip-container"
    >
      <div
        class="label-value__label"
        data-cy="active-analysts-label"
      >
        Active Analysts: 
      </div>
      <div
        class="label-value__value"
        data-cy="active-analysts-value"
        title="Active Analysts"
      >
        analyst 1, analyst 2
      </div>
    </div>
    <div
      class="label-value-container workflow-tooltip-container"
    >
      <div
        class="label-value__label"
        data-cy="start-time-label"
      >
        Start Time: 
      </div>
      <div
        class="label-value__value label-value__value--numeric"
        data-cy="start-time-value"
        title="Start Time"
      >
        500
      </div>
    </div>
    <div
      class="label-value-container workflow-tooltip-container"
    >
      <div
        class="label-value__label"
        data-cy="end-time-label"
      >
        End Time: 
      </div>
      <div
        class="label-value__value label-value__value--numeric"
        data-cy="end-time-value"
        title="End Time"
      >
        600
      </div>
    </div>
    <div
      class="label-value-container workflow-tooltip-container"
    >
      <div
        class="label-value__label"
        data-cy="last-modified-label"
      >
        Last Modified: 
      </div>
      <div
        class="label-value__value label-value__value--numeric"
        data-cy="last-modified-value"
        title="Last Modified"
      >
        1970-01-01 00:00:00
      </div>
    </div>
  </div>
</div>
`;

exports[`Workflow Tooltip matches tooltip snapshot 1`] = `
<div>
  <div
    class="bp4-popover-wrapper core-tooltip"
  >
    <div
      aria-haspopup="true"
      class="bp4-popover-target core-tooltip__target"
    >
      <span
        class=""
        tabindex="0"
      />
    </div>
  </div>
</div>
`;

exports[`Workflow Tooltip matches tooltip snapshot 2`] = `
<div>
  <div
    class="bp4-popover-wrapper core-tooltip"
  >
    <div
      aria-haspopup="true"
      class="bp4-popover-target core-tooltip__target"
    >
      <div
        class=""
        tabindex="0"
      >
        content
      </div>
    </div>
  </div>
</div>
`;
